components/nimbus/src/stateful/targeting.rs

```diff
@@ -27,3 +36,82 @@ impl NimbusTargetingHelper {
         }
     }
 }
+
+pub trait RecordedContext: Send + Sync {
+    /// Returns a JSON representation of the context object
+    ///
+    /// This method will be implemented in foreign code, i.e: Kotlin, Swift, Python, etc...
+    fn to_json(&self) -> JsonObject;
+
+    /// Returns a HashMap representation of the event queries that will be used in the targeting
+    /// context
+    ///
+    /// This method will be implemented in foreign code, i.e: Kotlin, Swift, Python, etc...
+    fn get_event_queries(&self) -> JsonObject;
+
+    /// Sets the object's internal value for the event query values
+    ///
+    /// This method will be implemented in foreign code, i.e: Kotlin, Swift, Python, etc...
+    fn set_event_query_values(&self, json: JsonObject);
+
+    /// Records the context object to Glean
+    ///
+    /// This method will be implemented in foreign code, i.e: Kotlin, Swift, Python, etc...
+    fn record(&self);
+}
+
+impl dyn RecordedContext {
+    pub fn execute_queries(
+        &self,
+        nimbus_targeting_helper: &NimbusTargetingHelper,
+    ) -> Result<Map<String, Value>> {
+        let results = Map::from_iter(self.get_event_queries().iter().filter_map(
+            |(key, value): (&String, &Value)| match value.as_str() {
+                Some(v) => match nimbus_targeting_helper.evaluate_jexl_raw_value(v.into()) {
+                    Ok(result) => Some((key.to_string(), result)),
+                    Err(err) => {
+                        log::info!(
+                            "error during jexl evaluation for query {} — {}",
+                            value,
+                            err.to_string()
+                        );
+                        None
+                    }
+                },
+                None => {
+                    log::info!("value {} for key {} is not a string", key, value);
+                    Some((key.to_string(), json!(value)))
+                }
```

Comment:
imo if they add non-string queries we should omit them.

Also will this logging ever be seen anywhere?

Reply:
Yes — it'll be seen in test runs now thanks to the TestLogger, and will be visible in Kotlin logs as well

Reply:
Actually, on the "None" section, we do need to return the value there because if its values are set more than once then it'll keep the value as it was already applied rather than removing it. That happens in tests and I suspect it'll happen in fenix because of how the messaging infra is set up